torch/csrc/distributed/rpc/request_callback_impl.cpp

```diff
@@ -212,14 +212,31 @@ std::shared_ptr<FutureMessage> RequestCallbackImpl::processRpc(
 
       // Process the original RPC.
       auto wrappedMessageType = rpcWithAutograd.wrappedMessageType();
-      auto wrappedRpcResponse = processRpc(
+      // Kick off processing for the nested future and get a Future<T> to the
+      // result.
+      auto wrappedRpcResponseFuture = processRpc(
           rpcWithAutograd.wrappedRpc(), wrappedMessageType, messageId);
-      wrappedRpcResponse->waitNoThrow(); // TODO: make async
 
-      return wrap(getMessageWithAutograd(
-          rpcWithAutograd.fromWorkerId(),
-          std::move(*wrappedRpcResponse).moveValue(),
-          MessageType::FORWARD_AUTOGRAD_RESP));
+      // Make an overall future for the wrapped response.
+      auto responseFuture = std::make_shared<rpc::FutureMessage>();
+      auto fromWorkerId = rpcWithAutograd.fromWorkerId();
+      // The original future needs to be marked as completed when the wrapped
+      // one completes, with the autograd context information wrapped.
+      wrappedRpcResponseFuture->addCallback(
+          [responseFuture, messageId, fromWorkerId, wrappedRpcResponseFuture](
+              const Message& m,
+              const c10::optional<utils::FutureError>& error) {
+            if (error) {
+              // Propagate error to responseFuture if we had one.
+              responseFuture->setError(error->what());
+            } else {
+              responseFuture->markCompleted(getMessageWithAutograd(
```

Comment:
I think you probably want to setId(messageId) on the message before marking completed with it.

Reply:
@jjlilley I'll add this, but could you clarify the reasoning why we need to do this? I couldn't find where the messageId was previously set when returning a wrapped message with `getMessageWithAutograd`, so I'm not sure what changes here would require us to set it.

Reply:
I'm suggestion doing so, simply because we were calling setId() previously, via the wrap() lambda call.

(If you go back to December, before we returned a future from this method, there was a function that ensured that any Message had the id set).

I honestly don't know if it's strictly necessary to do so, but it seems safer to keep this aspect of behavior as-is, unless there's a reason to change this.

Reply:
Added in the latest change.